### PR TITLE
datetime: Unify parse and print APIs

### PR DESCRIPTION
**Pull request by vmg**
## Description

This PR unifies the parsing and formatting APIs in the `datetime` package so the same code is used for both parsing and formatting all time specifiers in the Vitess codebase.

This includes:

- [x] Parsing of `TIME`, `DATETIME` and `DATE` literals
- [x] Parsing of `TIME`, `DATETIME` and `DATE` textual types into native `evalengine` types 
- [x] Generation of new timestamps for SQL datetime APIs (`NOW`, `CURDATE`, etc...)
- [x] Conversion between `TIME`, `DATETIME` and `DATE` textual types and numeric types
- [x] Arbitrary formatting of dates using `FORMAT_DATE`

cc @dbussink 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/datetime/types.go [hunk 1]
```diff
@@ -43,154 +45,242 @@ func (t *Time) AppendFormat(b []byte, prec uint8) []byte {
...
+		return 1
+	}
+	s1, s2 := t.Second(), t2.Second()
+	if s1 < s2 {
+		return -1
+	}
+	if s1 > s2 {
+		return 1
+	}
+	return 0
```
Comment (dbussink) [inline go/mysql/datetime/types.go:131]: @vmg We're missing looking at the nanoseconds here.

File: go/vt/vtgate/evalengine/compare.go [hunk 2]
```diff
@@ -121,17 +121,17 @@ func compareNumeric(left, right eval) (int, error) {
 //   - https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html
 //   - https://dev.mysql.com/doc/refman/8.0/en/date-and-time-type-conversion.html
 func compareDates(l, r *evalTemporal) (int, error) {
-	return compareGoTimes(l.toStdTime(), r.toStdTime())
+	return compareGoTimes(l.toStdTime(time.Local), r.toStdTime(time.Local))
```
Comment (dbussink) [inline go/vt/vtgate/evalengine/compare.go:124]: I think we can avoid having to convert to `time.Time` here and use `DateTime#Compare` here?

File: go/vt/vtgate/evalengine/compare.go [hunk 3]
```diff
@@ -121,17 +121,17 @@ func compareNumeric(left, right eval) (int, error) {
...
-		t2 = r.(*evalBytes).toDateBestEffort()
+		t1 = tt.toStdTime(time.Local)
+		t2 = r.(*evalBytes).toDateBestEffort().ToStdTime(time.Local)
 	} else if tt, ok := r.(*evalTemporal); ok {
-		t1 = l.(*evalBytes).toDateBestEffort()
-		t2 = tt.toStdTime()
+		t1 = l.(*evalBytes).toDateBestEffort().ToStdTime(time.Local)
+		t2 = tt.toStdTime(time.Local)
 	}
 	return compareGoTimes(t1, t2)
```
Comment (dbussink) [inline go/vt/vtgate/evalengine/compare.go:136]: Similar here I think.

**Review by dbussink — APPROVED** [review]
I think we should get this merged at this point before it gets really too unwieldy. It's already big and we have a bunch more but we can separate that.

File: go/mysql/datetime/format_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package datetime
+
```
Comment (frouioui) [inline go/mysql/datetime/format_test.go:2]: We're missing the header

File: go/mysql/datetime/types.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
+package datetime
+
+import (
```
Comment (frouioui) [inline go/mysql/datetime/types.go:3]: Here as well

**Review by frouioui — APPROVED** [review]

File: go/mysql/datetime/helpers.go [hunk 6]
```diff
@@ -67,3 +71,194 @@ func appendInt(b []byte, x int, width int) []byte {
...
+			c2 |= 'a' - 'A'
+			if c1 != c2 || c1 < 'a' || c1 > 'z' {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func lookup(tab []string, val string) (int, string, bool) {
```
Comment (systay) [inline go/mysql/datetime/helpers.go:93]:
```go
// lookup searches for the first matching substring in the 'tab' slice and compares it with the 'val' string.
// If a match is found, it returns the index of the matching substring in 'tab', the unmatched portion of 'val',
// and a boolean value 'true'.
// If no match is found, it returns -1, the original 'val' string, and a boolean value 'false'.
```

**Pull request merged**